assumeutxo: background validation completion

jamesob opened the pull request:
This is part of the [assumeutxo project](https://github.com/bitcoin/bitcoin/projects/11) (parent PR: https://github.com/bitcoin/bitcoin/pull/15606)

Part two of replacing https://github.com/bitcoin/bitcoin/pull/24232.

---

When a user activates a snapshot, the serialized UTXO set data is used to create an "assumed-valid" chainstate, which becomes active in an attempt to get the node to network tip as quickly as possible. Simultaneously in the background, the already-existing chainstate continues "conventional" IBD to both accumulate full block data and serve as a belt-and-suspenders to validate the assumed-valid chainstate.

Once the background chainstate's tip reaches the base block of the snapshot used, we set `m_stop_use` on that chainstate and immediately take the hash of its UTXO set; we verify that this matches the assumeutxo value in the source code. Note that while we ultimately want to remove this background chainstate, we don't do so until the following initialization process, when we again check the UTXO set hash of the background chainstate, and if it continues to match, we remove the (now unnecessary) background chainstate, and move the (previously) assumed-valid chainstate into its place. We then reinitialize the chainstate in the normal way.

As noted in previous comments, we could do the filesystem operations "inline" immediately when the background validation completes, but that's basically just an optimization that saves disk space until the next restart. It didn't strike me as worth the risk of moving chainstate data around on disk during runtime of the node, though maybe my concerns are overblown.

The final result of this completion process is a fully-validated chain, where the only evidence that the user synced using assumeutxo is the existence of a `base_blockhash` file in the `chainstate` directory.

jamesob committed:
add ChainstateManager.GetSnapshot{BaseHeight,BaseBlock}()

For use in later commits.

jamesob committed:
validation: add CChainState::m_disabled and ChainMan::isUsable

and remove m_snapshot_validated. This state can now be inferred by the
number of isUsable chainstates.

m_disabled is used to signal that a chainstate should no longer be used
by validation logic; it is used as a sentinel when background validation
completes or if the snapshot chainstate is found to be invalid.

isUsable is a convenience method that incorporates m_disabled.

jamesob committed:
add Chainstate::HasCoinsViews()

Used in subsequent commits. Also cleans up asserts in
coins_views-related convenience methods to be more exact.

jamesob committed:
docs: update assumeutxo.md

Include notes about the `chainstate_snapshot` rename as well as
updates for the included code.

jamesob committed:
validation: add ChainMan logic for completing UTXO snapshot validation

Trigger completion when a background validation chainstate reaches the
same height as a UTXO snapshot, and handle cleaning up the chainstate
on subsequent startup.

jamesob committed:
log: add LoadBlockIndex() message for assumedvalid blocks

I found this useful during unittest debugging.

jamesob committed:
test: add snapshot completion unittests

Also adjusts the previous snapshot chainstate init tests
to account for the fact that the init process is now attempting to
validate and complete background chainstates whose tip is at the
snapshot base block. We use a DisconnectTip() hack to preserve the
nature of the test.

jonatack commented:
```
test/validation_chainstatemanager_tests.cpp:418:35: error: calling function 'DisconnectTip' requires holding mutex 'bg_chainstate.m_mempool->cs' exclusively [-Werror,-Wthread-safety-precise]
        BOOST_CHECK(bg_chainstate.DisconnectTip(unused_state, &unused_pool));
                                  ^
test/validation_chainstatemanager_tests.cpp:418:35: note: found near match 'mempool->cs'
1 error generated.
```

ryanofsky commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -387,82 +387,57 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_activate_snapshot, SnapshotTestSetup)
-//!   chainstate only contains fully validated blocks and the other chainstate contains all blocks,
-//!   even those assumed-valid.
-//!
-BOOST_FIXTURE_TEST_CASE(chainstatemanager_loadblockindex, TestChain100Setup)
```
In commit "test: add testcases for snapshot initialization" (5ff0d2f7e78a40225a1eb9aa7bbf5c358d94bf48)

Something wonky is going on in this commit. It seems like some rebase error. This is dropping the `chainstatemanager_loadblockindex` test, and replacing it with an exact copy of the `chainstatemanager_snapshot_init` below, which now appears twice (and is edited in later commit 65ad0860ffd9441b005ac4bc240f596d9a81836f to become a different chainstatemanager_snapshot_completion test)

jamesob replied:
@ryanofsky yeah, thanks - not sure what happened there. Should be fixed.

ryanofsky reviewed (commented):
Code review ACK d4fc4ae18c6ac4e932beed20d6a61a018a5577d0 if missing chainstatemanager_loadblockindex test is added back. First 12 commits were previously reviewed as part of #25667. Later commits were cherry-picked from #24232 without code changes

jamesob commented:
Clang failure on some platforms looks spurious:

```
test/validation_chainstatemanager_tests.cpp:501:35: error: calling function 'DisconnectTip' requires holding mutex 'bg_chainstate.m_mempool->cs' exclusively [-Werror,-Wthread-safety-precise]
        BOOST_CHECK(bg_chainstate.DisconnectTip(unused_state, &unused_pool));
                                  ^
test/validation_chainstatemanager_tests.cpp:501:35: note: found near match 'mempool->cs'
1 error generated.
```
but lock is taken [here](https://github.com/bitcoin/bitcoin/pull/25740/commits/a9e27d3f242361534a1a50e403ab176b556ba528#diff-dbada1fe3a3d0af884304dd28be8c9df74b592401dec2c6400f6b491aefe6c9bR500). 

Anyone have ideas for workarounds?

ryanofsky commented:
If the lock is for `bg_chainstate.GetMempool()->cs` but the annotation requires `bg_chainstate.m_mempool->cs`, that compiler probably isn't going to know that `GetMempool()` and `m_mempool` are the same. Probably should stick with `m_mempool`

jamesob commented:
>  Probably should stick with m_mempool

It's protected :grimacing:. But looks like I can use the `RETURNS_LOCK` annotation on `MempoolMutex()`.

jamesob committed:
refactor: make MempoolMutex() public

for use in the following unittests.

jamesob commented:
Reopening and closing to get Github to deduplicate commits now included on master.

aureleoules commented:
@jamesob this PR should probably be added to #25740 now that it's out of draft.

jamesob commented:
Thanks for the ping @aureleoules; the parent PR description has been updated.

aureleoules commented on src/validation.cpp:
```diff
@@ -5117,6 +5168,146 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+       return SnapshotCompletionResult::SKIPPED;
+    }
+    int snapshot_tip_height = this->ActiveHeight();
+    int snapshot_base_height = *this->GetSnapshotBaseHeight();
```
a65af66bd5a292d0bd0030f0c6b49a3dfab0afcc: maybe `Assert` first?

aureleoules commented on src/validation.cpp:
```diff
@@ -5117,6 +5168,146 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+    }
+    int snapshot_tip_height = this->ActiveHeight();
+    int snapshot_base_height = *this->GetSnapshotBaseHeight();
+    CBlockIndex& index_new = *m_ibd_chainstate->m_chain.Tip();
```
a65af66bd5a292d0bd0030f0c6b49a3dfab0afcc: `Assert` here as well and can be const

aureleoules commented on src/validation.cpp:
```diff
@@ -5117,6 +5168,146 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+        return SnapshotCompletionResult::SKIPPED;
+    }
+
+    assert(SnapshotBlockhash());
+    uint256 snapshot_blockhash = *SnapshotBlockhash();
```
nit a65af66bd5a292d0bd0030f0c6b49a3dfab0afcc: can be inlined
```suggestion
	uint256 snapshot_blockhash{*Assert(SnapshotBlockhash())};

aureleoules commented on src/validation.cpp:
```diff
@@ -5121,6 +5170,146 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+        handle_invalid_snapshot();
+        return SnapshotCompletionResult::STATS_FAILED;
+    }
+    const auto ibd_stats = *maybe_ibd_stats;
```
a65af66bd5a292d0bd0030f0c6b49a3dfab0afcc
```suggestion
    const auto& ibd_stats = *maybe_ibd_stats;
```

aureleoules commented on src/validation.cpp:
```diff
@@ -5211,3 +5405,136 @@ Chainstate& ChainstateManager::ActivateExistingSnapshot(CTxMemPool* mempool, uin
+        return false;
+    }
+
+    auto snapshot_chainstate_path = *snapshot_chainstate_path_maybe;
+    auto ibd_chainstate_path = *ibd_chainstate_path_maybe;
```
nit: a65af66bd5a292d0bd0030f0c6b49a3dfab0afcc
```suggestion
    const auto& snapshot_chainstate_path = *snapshot_chainstate_path_maybe;
    const auto& ibd_chainstate_path = *ibd_chainstate_path_maybe;
```

aureleoules reviewed (commented):
I reviewed all the commits and verified they all compile. The code looks good and the unit tests seem to confirm that the behavior is correct.
I left some minor/nits comments.

jamesob commented on src/validation.cpp:
```diff
@@ -5117,6 +5168,146 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+          "snapshot base block %s (height %d). Snapshot is not valid.",
+          index_new.ToString(), snapshot_blockhash.ToString(), snapshot_base_height);
+       return SnapshotCompletionResult::BASE_BLOCKHASH_MISMATCH;
+    }
```
Good catch! Fixed.

jamesob commented on src/validation.cpp:
```diff
@@ -5213,6 +5404,39 @@ Chainstate& ChainstateManager::ActivateExistingSnapshot(CTxMemPool* mempool, uin
+    LogPrintf("[snapshot] renaming snapshot datadir %s to %s\n", dbpath, target);
+
+    try {
+        fs::rename(snapshot_datadir, invalid_path);
```
I think the idea is that we may want to preserve the bad snapshot datadir for later analysis or issue filing. I'll make a note in the abort node log message that the user can manually remove the directory if desired or save it for later forensics when filing an issue.

jamesob reviewed (commented):
Thanks very much for the reviews @achow101 @aureleoules. I've addressed all your feedback unless otherwise noted.

Just as a reminder, a previous (and very similar) version of this code was looked at by @ryanofsky @MarcoFalke back in https://github.com/bitcoin/bitcoin/pull/24232. Hopefully the code here is pretty familiar, and should make review a bit easier.

jamesob commented on src/validation.h:
```diff
@@ -570,12 +570,14 @@ class Chainstate
     //! Destructs all objects related to accessing the UTXO set.
     void ResetCoinsViews() { m_coins_views.reset(); }
 
+    bool HasCoinsViews() { return (bool)m_coins_views; }
```
Yep, done.

jamesob replied:
Ugh oops, my mistake. Fixed.

ariard commented on src/node/chainstate.cpp:
```diff
@@ -167,6 +134,93 @@ ChainstateLoadResult LoadChainstate(ChainstateManager& chainman, const CacheSize
+                                    const ChainstateLoadOptions& options)
+{
+    if (!hashAssumeValid.IsNull()) {
+        LogPrintf("Assuming ancestors of block %s have valid signatures.\n", hashAssumeValid.GetHex());
```
I think could do a `LookupBlockIndex` to display the block height, as it's more speaking for the node operator.

jamesob replied:
Out of scope for this PR, since this is just code I moved.

ariard commented on src/node/chainstate.cpp:
```diff
@@ -167,6 +134,93 @@ ChainstateLoadResult LoadChainstate(ChainstateManager& chainman, const CacheSize
+    auto [init_status, init_error] = CompleteChainstateInitialization(chainman, options);
+    if (init_status != ChainstateLoadStatus::SUCCESS) {
+        return {init_status, init_error};
+    }
```
I think you could have a quick log of the chainstate initialization success, as this function has a lot of checks/ops and you can inform the node operator where he is in the initialization sequence.

jamesob replied:
Out of scope or possibly follow-up given this was existing code.

ariard commented on src/validation.cpp:
```diff
@@ -2973,6 +2985,13 @@ bool Chainstate::ActivateBestChain(BlockValidationState& state, std::shared_ptr<
+    // Belt-and-suspenders check that we aren't attempting to advance the background
+    // chainstate past the snapshot base block.
+    if (WITH_LOCK(::cs_main, return m_disabled)) {
+        LogPrintf("m_disabled is set - this chainstate should not be in operation. Bug?.\n");
```
I think you could go even further, and invite the node operator to fulfill an issue on the repository or at least indicate a communication process to feedback the bug to us ?

jamesob replied:
Fixed.

ariard reviewed (commented):
Will do a more in-depth review soon

ryanofsky commented on src/validation.cpp:
```diff
@@ -5117,6 +5168,149 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+//      through IsUsable() checks, or
+//
+//  (ii) giving each chainstate its own lock instead of using cs_main for everything.
+SnapshotCompletionResult ChainstateManager::MaybeCompleteSnapshotValidation(
```
It looks like there has been a rebase error and two previous commits were combined into one commit that doesn't have a valid commit description. The two previous commits were:

- "move-only-ish: init: factor out chainstate initialization" (aead772a24b1f7759cd756bf6534a70dc41ca080)
- "validation: add ChainMan logic for completing UTXO snapshot validation" (863e5f389e08cd2069e4834df82879ec39142aac)

and the new combined commit is 

 - "move-only-ish: init: factor out chainstate initialization" (07b9dea9d4e8de2ce22e8ba50172a4af85801931)

It would help review to split these up and separate moveonly changes from other changes

jamesob replied:
Wow, no idea how that got so screwed up. Will fix, thanks.

ariard reviewed (commented):
Reviewing commit by commit. So far until 65d2ce6.

-----------------------------

I think this is pending on a rebase :) Few more PRs and we should be good to tag assumeutxo 0.1. Let's keep going!

jamesob committed:
move-only-ish: init: factor out chainstate initialization

Moves chainstate initialization into its own function. This is
necessary to later support a more readable way of handling
background-validation chainstate cleanup during init, since the
chainstate initialization functions may need to be repeated after
moving leveldb filesystem content around.

This commit isn't strictly necessary, but the alternative is to (ab)use
the `while` loop in init.cpp with a `continue` on the basis of a
specific ChainstateLoadingError return value from LoadChainstate. Not
only is this harder to read, but it can't be unittested.

The approach here lets us consolidate background-validation cleanup to
LoadChainstate, and therefore exercise it within tests.

This commit is most easily reviewed with

  git diff --color-moved=dimmed-zebra
  --color-moved-ws=ignore-space-change

jamesob commented on src/validation.h:
```diff
@@ -570,12 +570,14 @@ class Chainstate
     //! Destructs all objects related to accessing the UTXO set.
     void ResetCoinsViews() { m_coins_views.reset(); }
 
+    bool HasCoinsViews() const { return (bool)m_coins_views; }
```
Added.

jamesob commented on src/validation.h:
```diff
@@ -470,6 +470,13 @@ class Chainstate
 
+    //! This toggle exists for use when doing background validation for UTXO
+    //! snapshots. It is set once the background validation chain reaches the
+    //! same height as the base of the snapshot, and signals that we should no
```
Added.

ryanofsky commented on src/validation.cpp:
```diff
@@ -2810,6 +2810,18 @@ bool Chainstate::ConnectTip(BlockValidationState& state, CBlockIndex* pindexNew,
+        // This call may set `m_disabled`, which is referenced immediately afterwards in
+        // ActivateBestChain.
+        //
+        // If this fails (i.e. if snapshot is invalid), we will have marked the
+        // snapshot chainstate as unusable, and will have removed its coinsdb from
+        // disk.
```
In commit "validation: add ChainMan logic for completing UTXO snapshot validation" (e17d51768bcf5f7791fdf7d0b2ee4b11cc64d922)

I think it's just renamed `_INVALID`, not removed. In any case, this detail doesn't really seem relevant here. It seems more relevant that the failure is handled internally so no need to check for errors from this function.

ryanofsky commented on src/validation.cpp:
```diff
@@ -2810,6 +2810,18 @@ bool Chainstate::ConnectTip(BlockValidationState& state, CBlockIndex* pindexNew,
+    // If we are the background validation chainstate, check to see if we are done
+    // validating the snapshot (i.e. our tip has reached the snapshot's base block).
+    if (this != &m_chainman.ActiveChainstate()) {
+        // This call may set `m_disabled`, which is referenced immediately afterwards in
+        // ActivateBestChain.
```
In commit "validation: add ChainMan logic for completing UTXO snapshot validation" (https://github.com/bitcoin/bitcoin/commit/e17d51768bcf5f7791fdf7d0b2ee4b11cc64d922)

Would add "so it will stop connecting new blocks" to make it clearer what m_disabled is used for.

ryanofsky reviewed (approved):
Code review ACK e7dc5e6e52811e2b03c053ab07fc25720e6f1d10 assuming question about IsSnapshotValidated doesn't point at a real problem. Changes since last review were rebasing, adding const keywords, renaming stop_use to disabled, improving invalid snapshot message, adding mempool mutex to tests.

jamesob commented on src/validation.h:
```diff
@@ -962,7 +976,10 @@ class ChainstateManager
-    bool IsSnapshotValidated() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main) { return m_snapshot_validated; }
+    bool IsSnapshotValidated() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
+    {
+        return m_snapshot_chainstate && m_ibd_chainstate && m_ibd_chainstate->m_disabled;
```
Ah, yeah, good catch! I moved too fast when addressing Antoine's feedback and forgot that setting `m_disabled` on the IBD chainstate _does_ indicate full snapshot validation. It's a little confusing because setting `m_disabled` on the _snapshot_ chainstate indicates that validation failed, but this definition of `IsSnapshotValidated()` is actually still correct, I just screwed up the comment.

ryanofsky replied:
Oh, that is a helpful clarification that m_disabled is set in different conditions on the different chainstates. It'd be useful to say in doxygen comment for `m_disabled` that it is is only set on background chainstate after it is validated and validation was successful, and only set on snapshot chainstate if it's invalid and needs to be deleted.

jamesob commented on src/node/chainstate.cpp:
```diff
@@ -28,38 +28,13 @@
-                                    const ChainstateLoadOptions& options)
+// Complete initialization of chainstates after the initial call has been made
+// to ChainstateManager::InitializeChainstate().
+static ChainstateLoadResult CompleteChainstateInitialization(
```
I *would* do this, but repeating the function signature for a static function for perpetuity just to save on a diff that's pretty easy to read with `--color-moved=dimmed-zebra --color-moved-ws=ignore-space-change` anyway doesn't seem like the right trade to me. Will leave this as-is.

jamesob commented on src/validation.h:
```diff
@@ -955,6 +990,18 @@ class ChainstateManager
+    //!
+    //! If the coins match (expected), then mark the validation chainstate for
+    //! deletion and continue using the snapshot chainstate as active.
+    //! Otherwise, revert to using the ibd chainstate and shutdown (TODO).
```
Good catch!

jamesob reviewed (commented):
Thanks as always for your good review, @ryanofsky. I've rebased on top of latest master to fix CI.

ryanofsky commented on src/validation.h:
```diff
@@ -451,6 +451,19 @@ class Chainstate
+    //! In the expected case, it is set once the background validation chain reaches the
+    //! same height as the base of the snapshot and its UTXO set is found to hash to
+    //! the expected assumeutxo value. It signals that we should no longer connect
+    //! blocks to the background chainstate. When set on the background validation
+    //! chainstate, it signifies that we have fully validated the snapshot chainstate.
```
In commit "validation: add CChainState::m_disabled and ChainMan::isUsable" (e07764318ad766ba27dfeafe1ca2345534f81dd5)

This last sentence is confusing because it is part of the "In the expected case" paragraph, but it is referring to an different unexpected case. It's also not very clear that this sentence is referring to a different chainstate object than the rest of the paragraph.

Would move this sentence to a separate paragraph and say something like "In the error case, if the background UTXO hash does not match the expected assumeutxo valid, m_disabled will be set to true on the snapshot chainstate (not the background validation chainstate) indicating that the snapshot chainstate is invalid."

ryanofsky reviewed (approved):
Code review ACK ab41e8b4093925dc339096c15b33fe8b4726f55a. Just rebase since last review, and some nice comment and error message improvements.

ariard commented on src/validation.cpp:
```diff
@@ -3032,6 +3040,13 @@ bool Chainstate::ActivateBestChain(BlockValidationState& state, std::shared_ptr<
     LOCK(m_chainstate_mutex);
 
+    // Belt-and-suspenders check that we aren't attempting to advance the background
+    // chainstate past the snapshot base block.
```
From my understanding there is a marginal case where this is not *only* a belt-and-suspender. When background chainstate reaches height parity with the snapshot chainstate, the background chainstate is torn down only in `LoadChainstate->ValidatedSnapshotCleanup` by the call to `ResetChainstates()` L5584. Before, ABC could be called as the background chainstate is still `ActiveChainstate()`, at least for the block of the connection sequence, before the remaining one hit the `m_disabled` L3106. If correct, the comment could be more informative.

jamesob replied:
For what it's worth, in future PRs I think this possibility is eliminated, since calls like `ActiveChainstate().ActivateBestChain()` in ProcessNewBlock are replaced with specific chainstate references.

ariard reviewed (approved):
ACK ab41e8b

The code could be better documented, with few improvements, and the unit test coverage enhanced, however this could be done in a follow-up. Reviewed the validation and cleanup of the background chainstate from hiting the block height parity with snapshot in `ConnectTip`, to flushing the data struct when the node is re-started in `AppInitMain()`.

achow101 commented:
ACK 30e113735ae3b1465d30b2125e7826a7a1527b99

The commit message of 287bb3f4ce41341aa0aea97b8176dc790889830d "validation: add CChainState::m_disabled and ChainMan::isUsable" still mentions `m_stop_use` rather than `m_disabled`

jamesob commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -518,10 +533,156 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_snapshot_init, SnapshotTestSetup)
+        LOCK(chainman_restarted.GetMutex());
+        BOOST_CHECK_EQUAL(chainman_restarted.GetAll().size(), 1);
+        BOOST_CHECK(!chainman_restarted.IsSnapshotActive());
+        BOOST_CHECK(!chainman_restarted.IsSnapshotValidated());
```
Fixed.

jamesob commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -518,10 +533,156 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_snapshot_init, SnapshotTestSetup)
+
+    // Test tampering with the IBD UTXO set with an extra coin to ensure it causes
+    // snapshot completion to fail.
+    CCoinsViewCache& ibd_coins = WITH_LOCK(::cs_main,
```
Good for a follow-up.

jamesob commented on src/validation.cpp:
```diff
@@ -5196,6 +5241,149 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
 }
 
+// Currently, this function holds cs_main for its duration, which could be for
+// multiple minutes due to the GetUTXOStats call. This hold is necessary
```
Fixed.

jamesob commented:
Rebased and reworded the commit message that @achow101 pointed out was out of date. Also fixed two minor items that @ariard noted.

In attempt to encourage re-ACKers, here is the diff (including rebase items):

```diff
diff <(git diff --no-color au.complete.10~9..au.complete.10) <(git diff --no-color au.complete.11~9..au.complete.11)
109c109
< index 60acb614b4..d36184fed7 100644
---
> index ba1024d22e..cd82d8743c 100644
138c138
< -    if (nPruneTarget == std::numeric_limits<uint64_t>::max()) {
---
> -    if (chainman.m_blockman.GetPruneTarget() == std::numeric_limits<uint64_t>::max()) {
140,141c140,141
< -    } else if (nPruneTarget) {
< -        LogPrintf("Prune configured to target %u MiB on disk for block and undo files.\n", nPruneTarget / 1024 / 1024);
---
> -    } else if (chainman.m_blockman.GetPruneTarget()) {
> -        LogPrintf("Prune configured to target %u MiB on disk for block and undo files.\n", chainman.m_blockman.GetPruneTarget() / 1024 / 1024);
187c187
< +    if (nPruneTarget == std::numeric_limits<uint64_t>::max()) {
---
> +    if (chainman.m_blockman.GetPruneTarget() == std::numeric_limits<uint64_t>::max()) {
189,190c189,190
< +    } else if (nPruneTarget) {
< +        LogPrintf("Prune configured to target %u MiB on disk for block and undo files.\n", nPruneTarget / 1024 / 1024);
---
> +    } else if (chainman.m_blockman.GetPruneTarget()) {
> +        LogPrintf("Prune configured to target %u MiB on disk for block and undo files.\n", chainman.m_blockman.GetPruneTarget() / 1024 / 1024);
257c257
< index 22b9af1201..fd6a9b4c2e 100644
---
> index 56867a584b..45d59b685c 100644
294c294
< @@ -518,10 +533,156 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_snapshot_init, SnapshotTestSetup)
---
> @@ -518,10 +533,160 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_snapshot_init, SnapshotTestSetup)
366a367,368
> +    const Chainstate& active_cs2 = chainman_restarted.ActiveChainstate();
> +
371a374,375
> +        BOOST_CHECK(active_cs2.m_coinstip_cache_size_bytes > tip_cache_before_complete);
> +        BOOST_CHECK(active_cs2.m_coinsdb_cache_size_bytes > db_cache_before_complete);
453c457
< index 76bea97341..0916827710 100644
---
> @@ -3067,6 +3075,14 @@ bool Chainstate::ActivateBestChain(BlockValidationState& state, std::shared_ptr<
478c482,483
< +        LogPrintf("m_disabled is set - this chainstate should not be in operation. Bug?.\n");
---
> +        LogPrintf("m_disabled is set - this chainstate should not be in operation. " /* Continued */
> +            "Please report this as a bug. %s\n", PACKAGE_BUGREPORT);
485c490
< @@ -3080,6 +3095,15 @@ bool Chainstate::ActivateBestChain(BlockValidationState& state, std::shared_ptr<
---


```

achow101 commented:
re-ACK a00d0e78510d79bc9fc57ec622aec98a65efa8c0

ariard reviewed (approved):
ACK a00d0e78.

Since last ACK ab41e8b4:
- move `nPruneTarget` usage to `BlockManager::GetPruneTarget()` in `LoadChainstate()`
- bug report call message (i.e `PACKAGE_BUGREPORT`) in `LogPrintf` in `ActivateBestChain()`
- move `GetUTXOStats` mention to `ComputeUTXOStats` in `MaybeCompleteSnapshotValidation` comment
- new `BOOST_CHECK`s in `chainstatemanager_snapshot_completion` about cache size bytes

Sjors commented:
I'd like to test this with `loadtxoutset`. I wonder if I just cherry-pick a few things from #15606 to do that myself, or a more rigorous rebase is needed...

jamesob commented:
Maintainers: what else would you like to see for a merge here?

jamesob commented:
> In the pull description, could drop or update "Since it's dependent on the commits in #25667, I'm opening this as a draft."

Done, thanks.

brunoerg commented:
> The commit message of https://github.com/bitcoin/bitcoin/commit/287bb3f4ce41341aa0aea97b8176dc790889830d "validation: add CChainState::m_disabled and ChainMan::isUsable" still mentions m_stop_use rather than m_disabled

In the pull description, it still mention `m_stop_use`.

brunoerg reviewed (approved):
crACK a00d0e78510d79bc9fc57ec622aec98a65efa8c0

In 5ee22cdafd2562bcb8bf0ae6025e4b53c826382d, it creates `ChainstateManager::GetSnapshotBaseBlock` and  `ChainstateManager::GetSnapshotBaseHeight`. `GetSnapshotBaseHeight` is used (not in this commit) in `MaybeCompleteSnapshotValidation` to check if IBD (_in background_) has been completed.

In c29f26b47b8ef978d8689dc0222aa663361ee6cb, it creates `CChainState::m_disabled` and I could check it's been used to point that a chainstate should no longer be used by validation logic. (_perhaps PR description should be updated_).

Some logic I could understand:
![assumeutxo2](https://user-images.githubusercontent.com/19480819/222829606-2c2d943c-35a9-48f4-bccd-0cca56d62e10.jpg)


In e91b2165103784f13316e513d03dda0d63570675, it first added in `CompleteChainstateInitialization`
```cpp
assert(chainman.m_total_coinstip_cache > 0);
assert(chainman.m_total_coinsdb_cache > 0);
```
Good to check we have enough bytes available for in-memory coins caches and leveldb stuff.

In `LoadChainstate` I could check (what is mentioned):
1. Loads the fully validated chainstate
2. Loads a chain created from a UTXO snapshot, if any exist.

With both ones, it calls `MaybeCompleteSnapshotValidation` to compare the background chainstate's UTXO hash with the hard-coded (and expected) one. If so, it sets `m_disabled=true`. And then, we can _re-_initialize the chainstate (`InitializeChainstate`).

If we could have `loadtxoutset` (as mentioned by @Sjors), it would be good to review it and see things happen in practice.

achow101 commented:
Missed this on last review:

```
../../../src/validation.cpp: In member function ‘SnapshotCompletionResult ChainstateManager::MaybeCompleteSnapshotValidation(std::function<void(bilingual_str)>)’:
../../../src/validation.cpp:5450:14: error: catching polymorphic type ‘struct StopHashingException’ by value [-Werror=catch-value=]
 5450 |     } catch (StopHashingException) {
      |              ^~~~~~~~~~~~~~~~~~~~
```

jamesob commented:
> Missed this on last review (sorry):

Fixed!

achow101 commented:
ACK 2b373fe49d64f04ceab2309d3f40da7bac6b37d6

fjahr commented on src/validation.cpp:
```diff
@@ -5243,6 +5291,149 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+        return SnapshotCompletionResult::STATS_FAILED;
+    }
+
+    // XXX note that this function is slow and will hold cs_main for potentially minutes.
```
Just catching up... Not sure if intentional or not but maybe the XXX here was something you still wanted to address but missed @jamesob ?

maflcko reviewed (commented):
left some nits. Feel free to ignore
